File: sdk/typescript/examples/subscription-demo.ts
/**
 * Real-time Subscription Demo
 *
 * This example demonstrates the Circuit Breaker SDK's real-time subscription capabilities,
 * showing how to subscribe to various events including resource updates, workflow events,
 * agent executions, LLM streaming, cost updates, and MCP server status changes.
 */

import { Client } from '../src/client.js';
import {
  subscribeResourceUpdates,
  subscribeWorkflowEvents,
  subscribeLLMStream,
  subscribeCostUpdates,
  type ResourceUpdateEvent,
  type WorkflowEvent,
  type LLMStreamChunk,
  type CostUpdateEvent,
  type AgentExecutionEvent,
  type MCPServerStatusUpdate,
  type MCPSessionEvent,
  type SubscriptionId,
} from '../src/subscriptions.js';

async function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function main(): Promise<void> {
  console.log('🔔 Circuit Breaker Real-time Subscription Demo');
  console.log('===============================================');

  // Initialize the client
  const baseUrl = process.env.CIRCUIT_BREAKER_URL || 'http://localhost:4000';
  const apiKey = process.env.CIRCUIT_BREAKER_API_KEY;

  let client = Client.builder().baseUrl(baseUrl);
  if (apiKey) {
    client = client.apiKey(apiKey);
  }
  const circuitBreakerClient = client.build();

  // Test connection
  try {
    const ping = await circuitBreakerClient.ping();
    console.log(`✅ Connected to Circuit Breaker server: ${ping.message}`);
  } catch (error) {
    console.log(`❌ Failed to connect to server: ${error}`);
    console.log('   Note: This demo shows subscription infrastructure even without a running server');
    console.log(`   In production, ensure the Circuit Breaker server is running at ${baseUrl}`);
  }

  console.log('\n📡 Real-time Subscription Features:');
  console.log('===================================');

  // 1. Resource Updates Subscription
  console.log('\n1. 📊 Resource Updates Subscription');
  console.log('   ---------------------------------');

  const resourceId = 'demo_resource_123';
  console.log(`   Subscribing to updates for resource: ${resourceId}`);

  try {
    const resourceSubId = await circuitBreakerClient
      .subscriptions()
      .resourceUpdates()
      .resourceId(resourceId)
      .subscribe(
        (resource: ResourceUpdateEvent) => {
          console.log('   📦 Resource Update Received:');
          console.log(`      • ID: ${resource.id}`);
          console.log(`      • Workflow: ${resource.workflowId}`);
          console.log(`      • State: ${resource.state}`);
          console.log(`      • Updated: ${resource.updatedAt}`);
        },
        (error) => {
          console.log(`   ❌ Resource subscription error: ${error.message}`);
        },
        () => {
          console.log('   ✅ Resource subscription completed');
        },
      );

    console.log(`   ✅ Resource subscription active: ${resourceSubId.toString()}`);
  } catch (error) {
    console.log(`   ⚠️  Resource subscription setup: ${error}`);
  }

  // 2. Workflow Events Subscription
  console.log('\n2. 🔄 Workflow Events Subscription');
  console.log('   --------------------------------');

  const workflowId = 'demo_workflow_456';
  console.log(`   Subscribing to events for workflow: ${workflowId}`);

  try {
    const workflowSubId = await circuitBreakerClient
      .subscriptions()
      .workflowEvents()
      .workflowId(workflowId)
      .subscribe(
        (event: WorkflowEvent) => {
          console.log('   🔄 Workflow Event Received:');
          console.log(`      • ID: ${event.id}`);
          console.log(`      • Type: ${event.type}`);
          console.log(`      • Message: ${event.message}`);
          console.log(`      • Timestamp: ${event.timestamp}`);
        },
        (error) => {
          console.log(`   ❌ Workflow subscription error: ${error.message}`);
        },
        () => {
          console.log('   ✅ Workflow subscription completed');
        },
      );

    console.log(`   ✅ Workflow subscription active: ${workflowSubId.toString()}`);
  } catch (error) {
    console.log(`   ⚠️  Workflow subscription setup: ${error}`);
  }

  // 3. LLM Streaming Subscription
  console.log('\n3. 🤖 LLM Streaming Subscription');
  console.log('   ------------------------------');

  const requestId = 'llm_request_789';
  console.log(`   Subscribing to LLM stream for request: ${requestId}`);

  try {
    const llmSubId = await circuitBreakerClient
      .subscriptions()
      .llmStream(requestId)
      .subscribe(
        (chunk: LLMStreamChunk) => {
          console.log('   🤖 LLM Chunk Received:');
          console.log(`      • Request ID: ${chunk.id}`);
          console.log(`      • Content: ${chunk.content}`);
          console.log(`      • Finished: ${chunk.finished}`);
          console.log(`      • Timestamp: ${chunk.timestamp}`);
        },
        (error) => {
          console.log(`   ❌ LLM stream subscription error: ${error.message}`);
        },
        () => {
          console.log('   ✅ LLM stream subscription completed');
        },
      );

    console.log(`   ✅ LLM stream subscription active: ${llmSubId.toString()}`);
  } catch (error) {
    console.log(`   ⚠️  LLM stream subscription setup: ${error}`);
  }

  // 4. Cost Updates Subscription
  console.log('\n4. 💰 Cost Updates Subscription');
  console.log('   -----------------------------');

  const userId = 'demo_user_123';
  console.log(`   Subscribing to cost updates for user: ${userId}`);

  try {
    const costSubId = await circuitBreakerClient
      .subscriptions()
      .costUpdates()
      .userId(userId)
      .subscribe(
        (update: CostUpdateEvent) => {
          console.log('   💰 Cost Update Received:');
          if (update.userId) {
            console.log(`      • User ID: ${update.userId}`);
          }
          if (update.projectId) {
            console.log(`      • Project ID: ${update.projectId}`);
          }
          console.log(`      • Cost: $${update.cost.toFixed(2)}`);
          console.log(`      • Timestamp: ${update.timestamp}`);
        },
        (error) => {
          console.log(`   ❌ Cost updates subscription error: ${error.message}`);
        },
        () => {
          console.log('   ✅ Cost updates subscription completed');
        },
      );

    console.log(`   ✅ Cost updates subscription active: ${costSubId.toString()}`);
  } catch (error) {
    console.log(`   ⚠️  Cost updates subscription setup: ${error}`);
  }

  // 5. Agent Execution Stream
  console.log('\n5. 🤖 Agent Execution Stream');
  console.log('   --------------------------');

  const executionId = 'agent_exec_456';
  console.log(`   Subscribing to agent execution: ${executionId}`);

  try {
    const agentSubId = await circuitBreakerClient
      .subscriptions()
      .agentExecutionStream()
      .executionId(executionId)
      .subscribe(
        (event: AgentExecutionEvent) => {
          console.log('   🤖 Agent Execution Event:');
          console.log(`      • ID: ${event.id}`);
          console.log(`      • Agent: ${event.agentId}`);
          console.log(`      • Status: ${event.status}`);
          console.log(`      • Timestamp: ${event.timestamp}`);
        },
        (error) => {
          console.log(`   ❌ Agent execution subscription error: ${error.message}`);
        },
        () => {
          console.log('   ✅ Agent execution subscription completed');
        },
      );

    console.log(`   ✅ Agent execution subscription active: ${agentSubId.toString()}`);
  } catch (error) {
    console.log(`   ⚠️  Agent execution subscription setup: ${error}`);
  }

  // 6. MCP Server Status Updates
  console.log('\n6. 🔌 MCP Server Status Updates');
  console.log('   -----------------------------');

  const serverId = 'mcp_server_789';
  console.log(`   Subscribing to MCP server status: ${serverId}`);

  try {
    const mcpStatusSubId = await circuitBreakerClient
      .subscriptions()
      .mcpServerStatusUpdates()
      .serverId(serverId)
      .subscribe(
        (update: MCPServerStatusUpdate) => {
          console.log('   🔌 MCP Server Status Update:');
          console.log(`      • Server ID: ${update.serverId}`);
          console.log(`      • Status: ${update.status}`);
          if (update.message) {
            console.log(`      • Message: ${update.message}`);
          }
          console.log(`      • Timestamp: ${update.timestamp}`);
        },
        (error) => {
          console.log(`   ❌ MCP status subscription error: ${error.message}`);
        },
        () => {
          console.log('   ✅ MCP status subscription completed');
        },
      );

    console.log(`   ✅ MCP status subscription active: ${mcpStatusSubId.toString()}`);
  } catch (error) {
    console.log(`   ⚠️  MCP status subscription setup: ${error}`);
  }

  // 7. MCP Session Events
  console.log('\n7. 📡 MCP Session Events');
  console.log('   ----------------------');

  console.log(`   Subscribing to MCP session events for user: ${userId}`);

  try {
    const mcpSessionSubId = await circuitBreakerClient
      .subscriptions()
      .mcpSessionEvents()
      .userId(userId)
      .subscribe(
        (event: MCPSessionEvent) => {
          console.log('   📡 MCP Session Event:');
          console.log(`      • Session ID: ${event.sessionId}`);
          console.log(`      • Event: ${event.event}`);
          console.log(`      • Timestamp: ${event.timestamp}`);
        },
        (error) => {
          console.log(`   ❌ MCP session subscription error: ${error.message}`);
        },
        () => {
          console.log('   ✅ MCP session subscription completed');
        },
      );

    console.log(`   ✅ MCP session subscription active: ${mcpSessionSubId.toString()}`);
  } catch (error) {
    console.log(`   ⚠️  MCP session subscription setup: ${error}`);
  }

  // 8. Subscription Metrics and Monitoring
  console.log('\n8. 📈 Subscription Metrics');
  console.log('   ------------------------');

  const metrics = circuitBreakerClient.subscriptions().getMetrics();
  console.log('   Current Subscription Status:');
  console.log(`   • Active Subscriptions: ${metrics.activeSubscriptions}`);
  console.log(`   • Messages Received: ${metrics.messagesReceived}`);

  // 9. Convenience Functions Demo
  console.log('\n9. 🛠️  Convenience Functions');
  console.log('   -------------------------');

  console.log('   Using convenience functions for common subscriptions:');

  // Resource updates convenience function
  try {
    const convResourceSubId = await subscribeResourceUpdates(
      circuitBreakerClient,
      'convenience_resource',
      (resource) => {
        console.log(`   📦 Convenience resource update: ${resource.id} -> ${resource.state}`);
      },
    );

    console.log(`   ✅ Convenience resource subscription: ${convResourceSubId.toString()}`);
  } catch (error) {
    console.log(`   ⚠️  Convenience resource subscription: ${error}`);
  }

  // Workflow events convenience function
  try {
    const convWorkflowSubId = await subscribeWorkflowEvents(
      circuitBreakerClient,
      'convenience_workflow',
      (event) => {
        console.log(`   🔄 Convenience workflow event: ${event.type} - ${event.message}`);
      },
    );

    console.log(`   ✅ Convenience workflow subscription: ${convWorkflowSubId.toString()}`);
  } catch (error) {
    console.log(`   ⚠️  Convenience workflow subscription: ${error}`);
  }

  // LLM stream convenience function
  try {
    const convLLMSubId = await subscribeLLMStream(
      circuitBreakerClient,
      'convenience_llm_request',
      (chunk) => {
        console.log(`   🤖 Convenience LLM chunk: ${chunk.content} (finished: ${chunk.finished})`);
      },
    );

    console.log(`   ✅ Convenience LLM subscription: ${convLLMSubId.toString()}`);
  } catch (error) {
    console.log(`   ⚠️  Convenience LLM subscription: ${error}`);
  }

  // Cost updates convenience function
  try {
    const convCostSubId = await subscribeCostUpdates(
      circuitBreakerClient,
      'convenience_user',
      (update) => {
        console.log(`   💰 Convenience cost update: $${update.cost.toFixed(2)}`);
      },
    );

    console.log(`   ✅ Convenience cost subscription: ${convCostSubId.toString()}`);
  } catch (error) {
    console.log(`   ⚠️  Convenience cost subscription: ${error}`);
  }

  // 10. Advanced Subscription Patterns
  console.log('\n10. 🔬 Advanced Subscription Patterns');
  console.log('    ----------------------------------');

  // Multiple subscriptions for the same resource
  console.log('    Setting up multiple subscriptions for comprehensive monitoring:');

  const monitoredResourceId = 'monitored_resource_999';

  // State change monitoring
  try {
    const stateMonitorSubId = await circuitBreakerClient
      .subscriptions()
      .resourceUpdates()
      .resourceId(monitoredResourceId)
      .subscribe((resource) => {
        console.log(`    🔍 State Monitor: ${resource.id} is now in state '${resource.state}'`);
      });

    console.log(`    ✅ State monitor active: ${stateMonitorSubId.toString()}`);
  } catch (error) {
    console.log(`    ⚠️  State monitor setup: ${error}`);
  }

  // Workflow context monitoring
  try {
    const workflowMonitorSubId = await circuitBreakerClient
      .subscriptions()
      .workflowEvents()
      .workflowId('monitored_workflow')
      .subscribe((event) => {
        console.log(`    🔍 Workflow Monitor: ${event.message}`);
      });

    console.log(`    ✅ Workflow monitor active: ${workflowMonitorSubId.toString()}`);
  } catch (error) {
    console.log(`    ⚠️  Workflow monitor setup: ${error}`);
  }

  // 11. Real-time Dashboard Simulation
  console.log('\n11. 📊 Real-time Dashboard Simulation');
  console.log('    -----------------------------------');

  console.log('    Simulating a real-time dashboard with multiple data streams:');

  const dashboardMetrics = circuitBreakerClient.subscriptions().getMetrics();
  console.log('    📊 Dashboard Metrics:');
  console.log(`       • Total Active Streams: ${dashboardMetrics.activeSubscriptions}`);
  console.log(`       • Data Points Received: ${dashboardMetrics.messagesReceived}`);

  // Simulate dashboard updates
  for (let i = 1; i <= 5; i++) {
    await sleep(1000);
    const currentMetrics = circuitBreakerClient.subscriptions().getMetrics();
    console.log(
      `    📊 Dashboard Update #${i}: ${currentMetrics.activeSubscriptions} active streams, ${currentMetrics.messagesReceived} total messages`,
    );
  }

  // 12. TypeScript-specific Features
  console.log('\n12. 🔷 TypeScript-Specific Features');
  console.log('    --------------------------------');

  console.log('    Type-safe subscription handling:');

  // Type-safe event handlers with destructuring
  try {
    const typeSafeSubId = await circuitBreakerClient
      .subscriptions()
      .resourceUpdates()
      .resourceId('typescript_resource')
      .subscribe(
        ({ id, state, workflowId, updatedAt }: ResourceUpdateEvent) => {
          console.log('    🔷 Type-safe resource update:');
          console.log(`       • Resource: ${id}`);
          console.log(`       • State: ${state}`);
          console.log(`       • Workflow: ${workflowId}`);
          console.log(`       • Updated: ${updatedAt}`);
        },
        (error) => {
          // Type-safe error handling
          const errorMessage: string = error.message;
          const subscriptionId: string | undefined = error.subscriptionId?.toString();
          console.log(`    ❌ Type-safe error: ${errorMessage} (sub: ${subscriptionId})`);
        },
        () => {
          console.log('    ✅ Type-safe completion handler called');
        },
      );

    console.log(`    ✅ Type-safe subscription active: ${typeSafeSubId.toString()}`);
  } catch (error) {
    console.log(`    ⚠️  Type-safe subscription setup: ${error}`);
  }

  // Type-safe metrics access
  const typedMetrics = circuitBreakerClient.subscriptions().getMetrics();
  const activeCount: number = typedMetrics.activeSubscriptions;
  const messageCount: number = typedMetrics.messagesReceived;
  const failureCount: number = typedMetrics.connectionFailures;

  console.log('    🔷 Type-safe metrics access:');
  console.log(`       • Active (number): ${activeCount}`);
  console.log(`       • Messages (number): ${messageCount}`);
  console.log(`       • Failures (number): ${failureCount}`);

  // 13. Subscription Lifecycle Management
  console.log('\n13. 🔄 Subscription Lifecycle Management');
  console.log('    --------------------------------------');

  console.log('    Demonstrating subscription lifecycle:');
  console.log('    • All subscriptions are automatically managed');
  console.log('    • Auto-reconnection on connection loss');
  console.log('    • Graceful cleanup on application shutdown');
  console.log('    • Message queuing during disconnections');

  // Set up cleanup on process exit
  process.on('SIGINT', async () => {
    console.log('\n🔄 Cleaning up subscriptions...');
    try {
      await circuitBreakerClient.subscriptions().close();
      console.log('✅ All subscriptions closed gracefully');
      process.exit(0);
    } catch (error) {
      console.error('❌ Error during cleanup:', error);
      process.exit(1);
    }
  });

  // Wait a bit to show subscriptions are active
  console.log('\n⏰ Subscriptions are now active and listening for events...');
  console.log('   In a real application, this would continue running indefinitely.');
  console.log('   Events would be processed as they arrive from the server.');
  console.log('   Press Ctrl+C to gracefully shut down subscriptions.');

  await sleep(2000);

  // Final metrics
  const finalMetrics = circuitBreakerClient.subscriptions().getMetrics();
  console.log('\n📈 Final Subscription Statistics:');
  console.log(`   • Active Subscriptions: ${finalMetrics.activeSubscriptions}`);
  console.log(`   • Total Messages Processed: ${finalMetrics.messagesReceived}`);
  console.log(`   • Connection Failures: ${finalMetrics.connectionFailures}`);
  console.log(`   • Reconnection Attempts: ${finalMetrics.reconnectionAttempts}`);

  console.log('\n🎉 Subscription Demo Complete!');
  console.log('===============================');
  console.log('This demo showcased:');
  console.log('• Resource state change subscriptions');
  console.log('• Workflow event monitoring');
  console.log('• Real-time LLM streaming');
  console.log('• Cost update notifications');
  console.log('• Agent execution tracking');
  console.log('• MCP server status monitoring');
  console.log('• MCP session event handling');
  console.log('• Subscription metrics and monitoring');
  console.log('• Convenience functions for common patterns');
  console.log('• Advanced multi-stream monitoring');
  console.log('• Real-time dashboard simulation');
  console.log('• TypeScript-specific type safety features');
  console.log('• Subscription lifecycle management');
  console.log('\nThe Subscription infrastructure provides:');
  console.log('• Type-safe event handling with full TypeScript support');
  console.log('• Automatic reconnection and error recovery');
  console.log('• Comprehensive monitoring and metrics');
  console.log('• Builder patterns for ergonomic APIs');
  console.log('• Production-ready WebSocket management');
  console.log('• Real-time data streaming capabilities');
  console.log('• Graceful cleanup and resource management');

  // Keep the process running to demonstrate real-time capabilities
  console.log('\n⚡ Keeping process alive for real-time demonstration...');
  console.log('   Press Ctrl+C to exit gracefully');

  // Infinite loop to keep the process running
  // In a real app, this would be your main application logic
  while (true) {
    await sleep(5000);
    const liveMetrics = circuitBreakerClient.subscriptions().getMetrics();
    console.log(
      `📊 Live Metrics: ${liveMetrics.activeSubscriptions} subscriptions, ${liveMetrics.messagesReceived} messages`,
    );
  }
}

// Run the demo
main().catch((error) => {
  console.error('Demo failed:', error);
  process.exit(1);
});

File: sdk/rust/examples/subscription_demo.rs
//! Real-time Subscription Demo
//!
//! This example demonstrates the Circuit Breaker SDK's real-time subscription capabilities,
//! showing how to subscribe to various events including resource updates, workflow events,
//! agent executions, LLM streaming, cost updates, and MCP server status changes.

use circuit_breaker_sdk::{Client, Result};
use std::env;
use std::time::Duration;
use tokio;
use tokio::time::sleep;

#[tokio::main]
async fn main() -> Result<()> {
    println!("🔔 Circuit Breaker Real-time Subscription Demo");
    println!("===============================================");

    // Initialize the client
    let base_url =
        env::var("CIRCUIT_BREAKER_URL").unwrap_or_else(|_| "http://localhost:4000".to_string());
    let api_key = env::var("CIRCUIT_BREAKER_API_KEY").ok();

    let mut client_builder = Client::builder().base_url(&base_url)?;
    if let Some(key) = api_key {
        client_builder = client_builder.api_key(key);
    }
    let client = client_builder.build()?;

    // Test connection
    match client.ping().await {
        Ok(ping) => println!("✅ Connected to Circuit Breaker server: {}", ping.message),
        Err(e) => {
            println!("❌ Failed to connect to server: {}", e);
            println!("   Note: This demo shows subscription infrastructure even without a running server");
            println!(
                "   In production, ensure the Circuit Breaker server is running at {}",
                base_url
            );
        }
    }

    println!("\n📡 Real-time Subscription Features:");
    println!("===================================");

    // 1. Resource Updates Subscription
    println!("\n1. 📊 Resource Updates Subscription");
    println!("   ---------------------------------");

    let resource_id = "demo_resource_123";
    println!("   Subscribing to updates for resource: {}", resource_id);

    match client
        .subscriptions()
        .resource_updates()
        .resource_id(resource_id)
        .subscribe(|resource| {
            println!("   📦 Resource Update Received:");
            println!("      • ID: {}", resource.id);
            println!("      • Workflow: {}", resource.workflow_id);
            println!("      • State: {}", resource.state);
            println!("      • Updated: {}", resource.updated_at);
        })
        .await
    {
        Ok(subscription_id) => {
            println!(
                "   ✅ Resource subscription active: {}",
                subscription_id.to_string()
            );
        }
        Err(e) => {
            println!("   ⚠️  Resource subscription setup: {}", e);
        }
    }

    // 2. Workflow Events Subscription
    println!("\n2. 🔄 Workflow Events Subscription");
    println!("   --------------------------------");

    let workflow_id = "demo_workflow_456";
    println!("   Subscribing to events for workflow: {}", workflow_id);

    match client
        .subscriptions()
        .workflow_events()
        .workflow_id(workflow_id)
        .subscribe(|event| {
            println!("   🔄 Workflow Event Received:");
            println!("      • ID: {}", event.id);
            println!("      • Type: {}", event.event_type);
            println!("      • Message: {}", event.message);
            println!("      • Timestamp: {}", event.timestamp);
        })
        .await
    {
        Ok(subscription_id) => {
            println!(
                "   ✅ Workflow subscription active: {}",
                subscription_id.to_string()
            );
        }
        Err(e) => {
            println!("   ⚠️  Workflow subscription setup: {}", e);
        }
    }

    // 3. LLM Streaming Subscription
    println!("\n3. 🤖 LLM Streaming Subscription");
    println!("   ------------------------------");

    let request_id = "llm_request_789";
    println!("   Subscribing to LLM stream for request: {}", request_id);

    match client
        .subscriptions()
        .llm_stream(request_id)
        .subscribe(|chunk| {
            println!("   🤖 LLM Chunk Received:");
            println!("      • Request ID: {}", chunk.id);
            println!("      • Content: {}", chunk.content);
            println!("      • Finished: {}", chunk.finished);
            println!("      • Timestamp: {}", chunk.timestamp);
        })
        .await
    {
        Ok(subscription_id) => {
            println!(
                "   ✅ LLM stream subscription active: {}",
                subscription_id.to_string()
            );
        }
        Err(e) => {
            println!("   ⚠️  LLM stream subscription setup: {}", e);
        }
    }

    // 4. Cost Updates Subscription
    println!("\n4. 💰 Cost Updates Subscription");
    println!("   -----------------------------");

    let user_id = "demo_user_123";
    println!("   Subscribing to cost updates for user: {}", user_id);

    match client
        .subscriptions()
        .cost_updates()
        .subscribe(|update| {
            println!("   💰 Cost Update Received:");
            if let Some(uid) = &update.user_id {
                println!("      • User ID: {}", uid);
            }
            if let Some(pid) = &update.project_id {
                println!("      • Project ID: {}", pid);
            }
            println!("      • Cost: ${:.2}", update.cost);
            println!("      • Timestamp: {}", update.timestamp);
        })
        .await
    {
        Ok(subscription_id) => {
            println!(
                "   ✅ Cost updates subscription active: {}",
                subscription_id.to_string()
            );
        }
        Err(e) => {
            println!("   ⚠️  Cost updates subscription setup: {}", e);
        }
    }

    // 5. Agent Execution Stream
    println!("\n5. 🤖 Agent Execution Stream");
    println!("   --------------------------");

    let execution_id = "agent_exec_456";
    println!("   Subscribing to agent execution: {}", execution_id);

    match client
        .subscriptions()
        .agent_execution_stream()
        .subscribe(|event| {
            println!("   🤖 Agent Execution Event:");
            println!("      • ID: {}", event.id);
            println!("      • Agent: {}", event.agent_id);
            println!("      • Status: {}", event.status);
            println!("      • Timestamp: {}", event.timestamp);
        })
        .await
    {
        Ok(subscription_id) => {
            println!(
                "   ✅ Agent execution subscription active: {}",
                subscription_id.to_string()
            );
        }
        Err(e) => {
            println!("   ⚠️  Agent execution subscription setup: {}", e);
        }
    }

    // 6. MCP Server Status Updates
    println!("\n6. 🔌 MCP Server Status Updates");
    println!("   -----------------------------");

    let server_id = "mcp_server_789";
    println!("   Subscribing to MCP server status: {}", server_id);

    match client
        .subscriptions()
        .mcp_server_status_updates()
        .subscribe(|update| {
            println!("   🔌 MCP Server Status Update:");
            println!("      • Server ID: {}", update.server_id);
            println!("      • Status: {}", update.status);
            if let Some(msg) = &update.message {
                println!("      • Message: {}", msg);
            }
            println!("      • Timestamp: {}", update.timestamp);
        })
        .await
    {
        Ok(subscription_id) => {
            println!(
                "   ✅ MCP status subscription active: {}",
                subscription_id.to_string()
            );
        }
        Err(e) => {
            println!("   ⚠️  MCP status subscription setup: {}", e);
        }
    }

    // 7. MCP Session Events
    println!("\n7. 📡 MCP Session Events");
    println!("   ----------------------");

    println!("   Subscribing to MCP session events for user: {}", user_id);

    match client
        .subscriptions()
        .mcp_session_events()
        .subscribe(|event| {
            println!("   📡 MCP Session Event:");
            println!("      • Session ID: {}", event.session_id);
            println!("      • Event: {}", event.event);
            println!("      • Timestamp: {}", event.timestamp);
        })
        .await
    {
        Ok(subscription_id) => {
            println!(
                "   ✅ MCP session subscription active: {}",
                subscription_id.to_string()
            );
        }
        Err(e) => {
            println!("   ⚠️  MCP session subscription setup: {}", e);
        }
    }

    // 8. Subscription Metrics and Monitoring
    println!("\n8. 📈 Subscription Metrics");
    println!("   ------------------------");

    let metrics = client.subscriptions().manager().metrics();
    println!("   Current Subscription Status:");
    println!("   • Active Subscriptions: {}", metrics.active_count());
    println!("   • Messages Received: {}", metrics.messages_count());

    // 9. Convenience Functions Demo
    println!("\n9. 🛠️  Convenience Functions");
    println!("   -------------------------");

    println!("   Using convenience functions for common subscriptions:");

    // Resource updates convenience function
    match circuit_breaker_sdk::subscribe_resource_updates(
        &client,
        "convenience_resource",
        |resource| {
            println!(
                "   📦 Convenience resource update: {} -> {}",
                resource.id, resource.state
            );
        },
    )
    .await
    {
        Ok(subscription_id) => {
            println!(
                "   ✅ Convenience resource subscription: {}",
                subscription_id.to_string()
            );
        }
        Err(e) => {
            println!("   ⚠️  Convenience resource subscription: {}", e);
        }
    }

    // Workflow events convenience function
    match circuit_breaker_sdk::subscribe_workflow_events(&client, "convenience_workflow", |event| {
        println!(
            "   🔄 Convenience workflow event: {} - {}",
            event.event_type, event.message
        );
    })
    .await
    {
        Ok(subscription_id) => {
            println!(
                "   ✅ Convenience workflow subscription: {}",
                subscription_id.to_string()
            );
        }
        Err(e) => {
            println!("   ⚠️  Convenience workflow subscription: {}", e);
        }
    }

    // 10. Advanced Subscription Patterns
    println!("\n10. 🔬 Advanced Subscription Patterns");
    println!("    ----------------------------------");

    // Multiple subscriptions for the same resource
    println!("    Setting up multiple subscriptions for comprehensive monitoring:");

    let resource_id = "monitored_resource_999";

    // State change monitoring
    match client
        .subscriptions()
        .resource_updates()
        .resource_id(resource_id)
        .subscribe(|resource| {
            println!(
                "    🔍 State Monitor: {} is now in state '{}'",
                resource.id, resource.state
            );
        })
        .await
    {
        Ok(subscription_id) => {
            println!(
                "    ✅ State monitor active: {}",
                subscription_id.to_string()
            );
        }
        Err(e) => {
            println!("    ⚠️  State monitor setup: {}", e);
        }
    }

    // Workflow context monitoring
    match client
        .subscriptions()
        .workflow_events()
        .workflow_id("monitored_workflow")
        .subscribe(|event| {
            println!("    🔍 Workflow Monitor: {}", event.message);
        })
        .await
    {
        Ok(subscription_id) => {
            println!(
                "    ✅ Workflow monitor active: {}",
                subscription_id.to_string()
            );
        }
        Err(e) => {
            println!("    ⚠️  Workflow monitor setup: {}", e);
        }
    }

    // 11. Real-time Dashboard Simulation
    println!("\n11. 📊 Real-time Dashboard Simulation");
    println!("    -----------------------------------");

    println!("    Simulating a real-time dashboard with multiple data streams:");

    let dashboard_metrics = client.subscriptions().manager().metrics();
    println!("    📊 Dashboard Metrics:");
    println!(
        "       • Total Active Streams: {}",
        dashboard_metrics.active_count()
    );
    println!(
        "       • Data Points Received: {}",
        dashboard_metrics.messages_count()
    );

    // Simulate dashboard updates
    for i in 1..=5 {
        sleep(Duration::from_secs(1)).await;
        let current_metrics = client.subscriptions().manager().metrics();
        println!(
            "    📊 Dashboard Update #{}: {} active streams, {} total messages",
            i,
            current_metrics.active_count(),
            current_metrics.messages_count()
        );
    }

    // 12. Subscription Lifecycle Management
    println!("\n12. 🔄 Subscription Lifecycle Management");
    println!("    --------------------------------------");

    println!("    Demonstrating subscription lifecycle:");
    println!("    • All subscriptions are automatically managed");
    println!("    • Auto-reconnection on connection loss");
    println!("    • Graceful cleanup on application shutdown");
    println!("    • Message queuing during disconnections");

    // Wait a bit to show subscriptions are active
    println!("\n⏰ Subscriptions are now active and listening for events...");
    println!("   In a real application, this would continue running indefinitely.");
    println!("   Events would be processed as they arrive from the server.");

    sleep(Duration::from_secs(2)).await;

    // Final metrics
    let final_metrics = client.subscriptions().manager().metrics();
    println!("\n📈 Final Subscription Statistics:");
    println!(
        "   • Active Subscriptions: {}",
        final_metrics.active_count()
    );
    println!(
        "   • Total Messages Processed: {}",
        final_metrics.messages_count()
    );

    println!("\n🎉 Subscription Demo Complete!");
    println!("===============================");
    println!("This demo showcased:");
    println!("• Resource state change subscriptions");
    println!("• Workflow event monitoring");
    println!("• Real-time LLM streaming");
    println!("• Cost update notifications");
    println!("• Agent execution tracking");
    println!("• MCP server status monitoring");
    println!("• MCP session event handling");
    println!("• Subscription metrics and monitoring");
    println!("• Convenience functions for common patterns");
    println!("• Advanced multi-stream monitoring");
    println!("• Real-time dashboard simulation");
    println!("• Subscription lifecycle management");
    println!("\nThe Subscription infrastructure provides:");
    println!("• Type-safe event handling");
    println!("• Automatic reconnection and error recovery");
    println!("• Comprehensive monitoring and metrics");
    println!("• Builder patterns for ergonomic APIs");
    println!("• Production-ready WebSocket management");
    println!("• Real-time data streaming capabilities");

    Ok(())
}

// Helper struct to demonstrate structured event data
#[derive(Debug)]
pub struct ResourceUpdate {
    pub id: String,
    pub workflow_id: String,
    pub state: String,
    pub updated_at: String,
}

#[derive(Debug)]
pub struct WorkflowEventData {
    pub id: String,
    pub event_type: String,
    pub message: String,
    pub timestamp: String,
}

#[derive(Debug)]
pub struct LLMStreamChunk {
    pub id: String,
    pub content: String,
    pub finished: bool,
    pub timestamp: String,
}

#[derive(Debug)]
pub struct CostUpdate {
    pub user_id: Option<String>,
    pub project_id: Option<String>,
    pub cost: f64,
    pub timestamp: String,
}

#[derive(Debug)]
pub struct AgentExecutionEvent {
    pub id: String,
    pub agent_id: String,
    pub status: String,
    pub timestamp: String,
}

#[derive(Debug)]
pub struct MCPServerStatusUpdate {
    pub server_id: String,
    pub status: String,
    pub message: Option<String>,
    pub timestamp: String,
}

#[derive(Debug)]
pub struct MCPSessionEvent {
    pub session_id: String,
    pub event: String,
    pub timestamp: String,
}
